explicit media toggle and small bug fixes

alexjgriffith opened the pull request:
## Summary
This PR closes #152  and extends the fix for #150 to all places where ids were being converted from numbers to strings. It also adds some small fixes to the behaviour of threads and fixes boosting and favouriting (unfavouriting and unboosting still don't work).

## Changes
1. We now kill the http get request buffer once JSON has been extracted. 
2. `mastodon-tl--as-string` was implemented and replaced any occurrence of `number-to-string` or `int-to-string`
3. Added variable `mastodon-tl--display-media-p`. By default it is `'t` but can be made a local buffer variable and set to `nil`. When `nil` rather than displaying the media it just provides a link `Media::<link>`
4. Fixed checking for faves and boosts, they should now render properly. The return from `json-read-from-string` for `nil` is `:json-false` which evaluates to `'t` in elisp.
5. Fixed the `format` string that gets printed when faving and boosting
6. Fixed `mastodon-tl--thread` updating and requesting and changed its behaviour such that it tries to open the original toot thread rather than the boosted thread. 
7. Added tests for both the new `mastodon-tl--as-string` function and the `mastodon-tl--toot-id` utility.

hdurer commented on lisp/mastodon-http.el:
```diff
@@ -100,13 +100,14 @@ Pass response buffer to CALLBACK function."
   "Make GET request to URL. Return JSON response vector."
   (let ((json-vector
-         (with-current-buffer (mastodon-http--get url)
+         (with-current-bufer (mastodon-http--get url)
```
Surely that's wrong.  Should stay `buffer`

hdurer commented on lisp/mastodon-http.el:
```diff
@@ -100,7 +100,7 @@ Pass response buffer to CALLBACK function."
   "Make GET request to URL. Return JSON response vector."
   (let ((json-vector
-         (with-current-bufer (mastodon-http--get url)
+         (with-current-buffer (mastodon-http--get url)
```
Why not squash this commit with the previous one to keep the history sane and readable?

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -293,11 +293,14 @@ also render the html"
+                                (mastodon-media--get-media-link-rendering
+                                 preview-url)
+                              (concat "Media::" preview-url ))))
+                        media-attachements "\n")))
```
Why do we re-add the ugly newlines between images?  That's only needed if `mastodon-tl--display-media-p` is nil, or not?

hdurer replied:
I still think we should not add newlines between images when we do show images...

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -293,11 +293,14 @@ also render the html"
+                                 preview-url)
+                              (concat "Media::" preview-url ))))
+                        media-attachements "\n")))
+    (if (not (equal media-string "")) 
```
use `string=` and only do this newline prepending if not displaying media.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -317,8 +320,7 @@ also render the html"
     (substring (mastodon-tl--content toot) 0 -2)
-    (when mastodon-tl--display-media-p
-      (mastodon-tl--media toot))
+    (mastodon-tl--media toot)
```
We have only just made this change and now it is being undone again.  Would it make sense to just squash this commit with the first one?

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -1,4 +1,4 @@
-;;; mastodon-tl.el --- HTTP request/response functions for mastodon.el  -*- lexical-binding: t -*-
```
a) all the files seem to use 3 semicolons, why change it in only one?
b) this seems unrelated to this PR.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -399,21 +399,35 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
   (mastodon-tl--property 'toot-id t))
 
+(defun mastodon-tl--as-string(id)
+  (cond ((numberp id )
```
Add doc string

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -399,21 +399,35 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
   (goto-char (point-max))
   (mastodon-tl--property 'toot-id t))
 
+(defun mastodon-tl--as-string(id)
```
`mastodon-tl--id-as-string`?

alexjgriffith replied:
This could be used anywhere we wish to convert a number to a string. To generalize I will rename the variable `ID` to `NUMERIC`.

hdurer replied:
ok.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -399,21 +399,35 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
     (with-output-to-temp-buffer buffer
       (switch-to-buffer buffer)
+      (mastodon-mode)
+      (setq mastodon-tl--buffer-spec
```
Nit: This change seems unrelated to what the commit message says...

alexjgriffith replied:
Updated the commit message to include a line addressing this move. I was having odd behaviour in the thread buffer if i defined mastodon-tl--buffer-spec` before initializing `mastodon-mode`.

hdurer commented on lisp/mastodon-toot.el:
```diff
@@ -99,7 +99,7 @@ Remove MARKER if RM is non-nil."
                            (lambda ()
                              (mastodon-toot--action-success "F" remove)
-                             (message (format "%sd #%d" action id))))))
+                             (message (format "%s #%d" action id))))))
```
Now that ids are strings, wouldn't the second formatting param also be `%s`?

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -408,10 +408,24 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
+  (let* ((json (mastodon-tl--property 'toot-json))
+         (id (alist-get 'id json ))
+         (reblog (alist-get 'reblog json )))
+    (if reblog
```
Should this be wrapped in the new "make it a string" helper function to ensure we return valid ids?

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -423,7 +437,7 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
                           endpoint ,(format "statuses/%s/context" id)
                           update-function
-                          mastodon-tl--timeline))
+                          (lambda(toot) (message "END of thread."))))
```
Nit: space after `lambda`

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -236,8 +235,8 @@ TIME-STAMP is assumed to be in the past."
         (parsed-time (date-to-time (mastodon-tl--field 'created_at toot)))
-        (faved (equal ":json-true" (mastodon-tl--field 'favourited toot)))
-        (boosted (equal ":json-true" (mastodon-tl--field 'reblogged toot))))
+        (faved (equal 't (mastodon-tl--field 'favourited toot)))
```
Whoops!  That looks like either our tests are wrong or some test is missing?

hdurer replied:
Ah ok, I now get it (reading the commit message).

Resolved.

hdurer reviewed (commented):
First pass through these changes.  I find it very hard to read and understand this PR.  I think it would be preferable if you squashed all this into 2 or 3 commits.

There is really no need to keep all the false starts, accidental changes and undoing of those accidental changes.

hdurer reviewed (dismissed):
Seems fine.  I suggest you still squash these two into one before merging into develop.

I still think adding newlines between images is wrong but I guess that's a matter of taste?

alexjgriffith committed:
Closes #152 and extends the fix for #150 as well as a host of bug fixes

We now kill the http get request buffer once JSON has been extracted.

mastodon-tl--as-string was implemented and replaced any occurrence of number-to-string or int-to-string

Added variable mastodon-tl--display-media-p. By default it is 't but can be made a local buffer variable and set to nil. When nil rather than displaying the media it just provides a link Media::<link>

Fixed checking for faves and boosts, they should now render properly. The return from json-read-from-string for nil is :json-false which evaluates to 't in elisp.

Fixed the format string that gets printed when faving and boosting

Fixed mastodon-tl--thread updating and requesting and changed its behaviour such that it tries to open the original toot thread rather than the boosted thread.

Added tests for both the new mastodon-tl--as-string function and the mastodon-tl--toot-id utility.

enter mastodon mode before defining local buffer variable mastodon-tl--buffer-spec. This fixes some oddities with the local buffer variable.

hdurer reviewed (approved):
LGTM
